Mathlib/CategoryTheory/CodiscreteCategory.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2024 Alvaro Belmonte. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alvaro Belmonte
+-/
+import Mathlib.CategoryTheory.EqToHom
+import Mathlib.CategoryTheory.Pi.Basic
+import Mathlib.Data.ULift
+import Mathlib.CategoryTheory.Category.Cat
+import Mathlib.CategoryTheory.Adjunction.Basic
+
+/-!
+# Codiscrete categories
+
+We define `Codiscrete A` as an alias for the type `A` ,
+and use this type alias to provide a `Category` instance
+whose Hom type are Unit types.
+
+`Codiscrete.lift` promotes a function `f : C → A` (for any category `C`) to a functor
+`Discrete.functor f : C ⥤  Codiscrete A`.
+
+Similarly, `Codiscrete.natTrans` and `Codiscrete.natIso` promote `I`-indexed families of morphisms,
+or `I`-indexed families of isomorphisms to natural transformations or natural isomorphism.
+
+We define `functor : Type u ⥤ Cat.{0,u}` which sends a type to the codiscrete category and show
+it is right adjoint to `Cat.objects.`
+-/
+namespace CategoryTheory
+
+universe u v w
+/-- We start by introducting an alias for the type underlying a codiscrete category
+structure.-/
+def Codiscrete (A : Type u) : Type u := A
+
+namespace Codiscrete
+
+
+instance (A : Type*) : Category (Codiscrete A) where
+  Hom _ _ := Unit -- The hom types in the Codiscrete A are the unit type.
+  id _ := ⟨⟩ -- This is the unique element of the unit type.
+  comp _ _ := ⟨⟩
+
+/-- A function induces a functor between codiscrete categories.-/
+def funToFunc {A B : Type*} (f : A → B) : Codiscrete A ⥤ Codiscrete B where
+  obj a := f a
+  map _ := ⟨⟩
+
+section
+variable {C : Type u} [Category.{v} C] {A : Type w}
+
+/-- Any function `C → A` lifts to a functor `C ⥤  Codiscrete A`. For discrete categories this is
+called `functor` but we use that name for something else. -/
+def lift (F : C → A) : C ⥤ Codiscrete A where
```

Comment:
I would suggest using the name `Codiscrete.functor` so as to keep a certain parallelism with `Discrete.functor`.

Reply:
@joelriou I agree with the virtue of parallelism but in this case I think "lift" is the better name and functor should refer to the thing that sends a function on types to a functor on codiscrete categories. How would you feel about keeping this terminology here and then changing the terminology used in the discrete categories file? 

Note the discrete file is missing the analogous adjunction. @AlvaroRBO was planning to add this in a separate PR, though if you'd prefer it could be done in this PR.

Reply:
I think we should not change `Discrete....` definitions because they follows a very good logic (`Discrete.functor`, `Discrete.natTrans`, etc.), and they perfectly abide the naming conventions (`functor` is a functor, `natTrans` is natural transformation, etc).
We may keep `Codiscrete.lift` if there is no better alternative.

Reply:
Sorry @joelriou I don't mean to be obtuse. One thing that the discrete file doesn't have is a name for the functor from Type.{u} to Cat.{0,u} that constructs the discrete category on a type and the functor between such from a function. What would you call this? (Or, more relevantly, what should @AlvaroRBO call the codiscrete analogue of this?) In other words, what names should be used for the left and right adjoints of `Cat.objects : Cat --> Type`?

Reply:
Currently, we have very few constructions involving `Cat`. In general, I would suggest putting definitions involving `Cat` in the `Cat` namespace, or writing `Cat` in the names if some definitions are put in another namespace, like `Discrete.functorToCat`. I would reserve short names like `functor`, `natTrans`, etc, for the most basic/lowlevel constructions.

Reply:
Got it. Thanks for the explanation. As a category theorist, I always default to thinking about Cat, but I can see how that's not the most useful perspective ;)